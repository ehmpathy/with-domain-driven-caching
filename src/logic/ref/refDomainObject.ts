type Constructor<T> = new (args: T) => T;

/**
 * define a reference to a property
 */
export const refDomainObject = <T extends Record<string, any>>(
  dobj: Constructor<T>,
): string => dobj.name;
