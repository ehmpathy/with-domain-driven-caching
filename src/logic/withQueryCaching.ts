import {
  normalizeDomainObjectReferences,
  withNormalization,
  withSerialization,
} from 'with-cache-normalization';
import { SerializableObject } from 'with-cache-normalization/dist/domain/NormalizeCacheValueMethod';
import { SimpleCache } from 'with-simple-caching';

import { DomainDrivenQueryDependsOn } from '../domain/DomainDrivenQueryDependsOn';
import { LogMethod } from '../domain/constants';
import { addQueryKeyToDependencyPointer } from './dependencyPointers/addQueryKeyToDependencyPointer';
import { getDependencyPointersDependedOnByQuery } from './dependencyPointers/getDependencyPointersDependedOnByQuery';

export const DEFAULT_REFERENCE_SECONDS_UNTIL_EXPIRATION = 30 * 24 * 60 * 60; // keep them for a long time, since otherwise we may invalidate queries before they should be

/**
 * adds domain driven caching to a domain.logic.query
 *
 * features
 * - ‚ú® automatic cache updates are enabled out of the box, due to [domain.object dereferencing](../../readme.md)
 * - ‚ú® automatic cache invalidation can be easily enabled by defining what your query dependsOn, due to [dependency pointers](../../readme.md)
 * - üî≠ observable tracing of underlying operations can be easily enabled by passing in the logMethod you'd like operations to be reported to
 *
 * note
 * - ‚ö†Ô∏è ensure all of the domain.object relationships and identities your domain.logic.query depends on are explicitly defined via the `dependsOn` attribute, to activate automated cache invalidation
 */
export const withQueryCaching = <I extends any[], O extends SerializableObject>(
  logic: (...args: I) => Promise<O>,
  options: {
    /**
     * the cache to use
     */
    cache: SimpleCache<string>;

    /**
     * the logical dependencies of your domain.logic.query
     *
     * note
     * - the dependencies listed here drive the automatic cache invalidation ‚ú®
     * - ensure all logical dependencies are listed to prevent stale data in the cache ‚ö†Ô∏è
     *
     * for example
     * - a query called `getPaymentByUuid({ uuid }) => Payment | null` will likely have a dependency on `{ identity: { dobj: Payment, uuid: ({ input }) => input[0].uuid } }`
     * - a query called `getLastPaymentOfUser({ userUuid }) => Payment` will likely have a dependency on `{ relationship: { from: { dobj: User, uuid: ({ input }) => input[0].uuid }, to: { dobj: Payment }, via: { dobj: Payment, prop: 'userUuid' } }`
     *
     * tips
     * - you can see the dependency pointers your query will be invalidated by, by searching for logs with the message `ddcache.query.set`. if `logDebug` is set, this log will be emitted each time your query is set to the cache
     */
    dependsOn: DomainDrivenQueryDependsOn<I, O>;

    /**
     * the serialization strategy to use
     */
    serialize?: {
      /**
       * the key serialization strategy to use
       *
       * note
       * - if you expect to use this with [simple-lambda-client](https://github.com/ehmpathy/simple-lambda-client), be sure to use it's `getSimpleLambdaClientCacheKey` method
       */
      key: (...args: I) => string;
    };

    /**
     * a method to define whether the the results are valid to cache
     *
     * for example
     * - invalidate null results from the start -> queries will hit underlying data store until result is found
     */
    valid?: ({ input, output }: { input: I; output: O }) => boolean;

    /**
     * a custom seconds until expiration to use for this query, if desired
     */
    secondsUntilExpiration?: number;

    /**
     * the log method via which to report debug logs, if observability is desired üî≠
     */
    logDebug?: LogMethod;
  },
): ((...args: I) => Promise<O>) => {
  // add domain.object reference normalization and serialization to the cache the user asked to use, for automatic cache updates
  const cache = withNormalization(
    withSerialization<SerializableObject>(options.cache),
    {
      normalize: normalizeDomainObjectReferences,
      referenceSecondsUntilExpiration:
        DEFAULT_REFERENCE_SECONDS_UNTIL_EXPIRATION,
    },
  );

  // define the query with caching
  return async (...input: I): Promise<O> => {
    // the key to use for the cache for this request
    const key = options.serialize?.key(...input) ?? JSON.stringify(input);

    // check whether the response is already the cache
    const found = await cache.get(key);
    if (found) {
      if (options.logDebug) options.logDebug('ddcache.query.get.hit', { key });
      return found as O; //if it is, we're donzo! no need to run the expensive logic
    }
    if (options.logDebug) options.logDebug('ddcache.query.get.miss', { key });

    // execute the logic to get the response
    const output = await logic(...input);

    // determine if the output is valid
    const isValidCacheEntry =
      !options.valid || options.valid({ input, output });
    if (!isValidCacheEntry) {
      if (options.logDebug)
        options.logDebug('ddcache.query.set.skip', {
          key,
          valid: isValidCacheEntry,
        });
      return output; // if its not a valid cache entry, don't save it to the cache
    }

    // set the dependency pointers for this query to the cache (do this first, to ensure the query isn't cached until its dependency pointers are, preventing missed invalidations in case of partial success)
    const pointers = getDependencyPointersDependedOnByQuery({
      dependsOn: options.dependsOn,
      execution: {
        input,
        output,
      },
    });
    await Promise.all(
      pointers.map((pointer) =>
        addQueryKeyToDependencyPointer({ cache, pointer, queryKey: key }),
      ),
    );

    // set the value to the cache
    await cache.set(key, output, {
      secondsUntilExpiration: options.secondsUntilExpiration,
    });
    if (options.logDebug)
      options.logDebug('ddcache.query.set', { key, deps: pointers });

    // return the cached value
    const foundNow = await cache.get(key);
    if (!foundNow)
      console.warn(
        'could not get value from cache immediately after setting. is something wrong? returning the value directly for now to not block operation',
        { key },
      );
    return foundNow as O;
  };
};
