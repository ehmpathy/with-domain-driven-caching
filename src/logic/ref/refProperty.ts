import { refDomainObject } from './refDomainObject';

type Constructor<T> = new (args: T) => T;

/**
 * define a reference to a property
 */
export const refProperty = <T extends Record<string, any>, P extends keyof T>(
  dobj: Constructor<T>,
  prop: P,
): { dobj: string; prop: string } => ({
  dobj: refDomainObject(dobj),
  prop: prop as string,
});
