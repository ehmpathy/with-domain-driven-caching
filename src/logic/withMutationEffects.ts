import {
  normalizeDomainObjectReferences,
  withSerialization,
} from 'with-cache-normalization';
import { SerializableObject } from 'with-cache-normalization/dist/domain/NormalizeCacheValueMethod';
import { SimpleCache } from 'with-simple-caching';

import { LogMethod } from '../domain/constants';
import { getDependencyPointersInvalidatedByMutation } from './dependencyPointers/getDependencyPointersInvalidatedByMutation';
import { invalidateQueriesByDependencyPointer } from './dependencyPointers/invalidateQueriesByDependencyPointer';
import { DEFAULT_REFERENCE_SECONDS_UNTIL_EXPIRATION } from './withQueryCaching';

/**
 * adds domain driven cache effects to a domain.logic.mutation
 *
 * features
 * - ‚ú® automatic cache updates are driven by the outputs of your mutation, via [domain.object dereferencing](../../readme.md)
 * - ‚ú® automatic cache invalidation are driven by the outputs of your mutation, via [dependency pointers](../../readme.md)
 * - üî≠ observable tracing of underlying operations can be easily enabled by passing in the logMethod you'd like operations to be reported to
 *
 * note
 * - ‚ö†Ô∏è ensure all of the domain.objects affected by your mutation are included in the output, to drive the automatic cache updates and invalidations
 */
export const withMutationEffects = <
  I extends any[],
  O extends SerializableObject,
>(
  logic: (...args: I) => Promise<O>,
  options: {
    /**
     * the cache to use
     */
    cache: SimpleCache<string>;

    /**
     * the log method via which to report debug logs, if observability is desired üî≠
     */
    logDebug?: LogMethod;
  },
): typeof logic => {
  // add domain.object reference normalization and serialization to the cache the user asked to use, for automatic cache updates
  const cache = withSerialization<SerializableObject>(options.cache);

  // execute the mutation with effects
  return async (
    ...input: Parameters<typeof logic>
  ): ReturnType<typeof logic> => {
    // execute the logic and get the response
    const output = await logic(...input);

    // trigger cache invalidations based on the response
    const pointers = await getDependencyPointersInvalidatedByMutation({
      cache,
      output,
    });
    const queriesInvalidated = (
      await Promise.all(
        pointers.map((pointer) =>
          invalidateQueriesByDependencyPointer({ cache, pointer }),
        ),
      )
    ).flat();

    // trigger cache updates by setting the normalized references to the cache
    const { references } = normalizeDomainObjectReferences({ value: output });
    await Promise.all(
      Object.entries(references).map(([refKey, refDobj]) =>
        cache.set(refKey, refDobj, {
          secondsUntilExpiration: DEFAULT_REFERENCE_SECONDS_UNTIL_EXPIRATION,
        }),
      ),
    );

    // log about the outcome
    if (options.logDebug)
      options.logDebug('ddcache.mutation.effects', {
        updated: {
          references: Object.keys(references),
        },
        evaluated: {
          deps: pointers,
        },
        invalidated: {
          queries: queriesInvalidated,
        },
      });

    // return the output
    return output;
  };
};
