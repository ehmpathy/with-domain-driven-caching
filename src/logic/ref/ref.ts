import { refDomainObject } from './refDomainObject';
import { refProperty } from './refProperty';

type Constructor<T> = new (args: T) => T;

/**
 * define a reference to a domain.object or domain.object.property
 *
 * for example
 * - `ref(Ship)` produces a reference to the `Ship` domain object
 * - `ref(Ship, 'isid')` produces a reference to the `Ship.isid` property
 *
 * note
 * - this method uses type overloads to support typing different outputs per inputs
 *   - if only `dobj` is passed in, refDomainObject is called, and it's return type is returned
 *   - if both `dobj` and `prop` are passed in, refProperty is called, and its return type is returned
 */
export function ref<T extends Record<string, any>>(
  dobj: Constructor<T>,
): ReturnType<typeof refDomainObject>;
export function ref<T extends Record<string, any>, P extends keyof T>(
  dobj: Constructor<T>,
  prop: P,
): ReturnType<typeof refProperty>;
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function ref<T extends Record<string, any>, P extends keyof T>(
  dobj: Constructor<T>,
  prop?: P,
) {
  if (prop) return refProperty(dobj, prop);
  return refDomainObject(dobj);
}
