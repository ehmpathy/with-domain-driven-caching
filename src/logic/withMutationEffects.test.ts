import { DomainEntity, DomainValueObject } from 'domain-objects';
import { SimpleAsyncCache } from 'with-simple-caching';

import { ref } from './ref/ref';
import { withMutationEffects } from './withMutationEffects';
import { withQueryCaching } from './withQueryCaching';

interface Ship {
  uuid: string;
  isid: string; // international ship id
  name: string;
}
class Ship extends DomainEntity<Ship> implements Ship {
  public static unique = ['isid'];
  public static updatable = ['name'];
}
interface ContainerLock {
  manufacturer: string;
  model: string;
}
class ContainerLock
  extends DomainValueObject<ContainerLock>
  implements ContainerLock {}

interface Container {
  uuid: string;
  icid: string; // international container id
  manufacturer: string;
  onShipUuid: string;
  insuredCargoUuids: string[];
  lock: ContainerLock;
}
class Container extends DomainEntity<Container> implements Container {
  public static unique = ['icid'];
  public static updatable = ['onShipUuid', 'insuredCargoUuids', 'lock'];
}

const lock = new ContainerLock({
  manufacturer: 'swiper no swiping',
  model: 'locky mclockface',
});

const getContainerAndShip = async ({
  containerUuid,
  shipUuid,
}: {
  containerUuid: string;
  shipUuid: string;
}) => {
  return {
    container: new Container({
      uuid: containerUuid,
      icid: '__container.icid__',
      manufacturer: 'containers r us',
      onShipUuid: shipUuid,
      insuredCargoUuids: ['__insured.cargo.1__', '__insured.cargo.2__'],
      lock,
    }),
    ship: new Ship({
      uuid: shipUuid,
      isid: '__ship.isid__',
      name: 'boaty mcboatface',
    }),
  };
};

const addContainerToShip = async ({
  containerUuid,
  shipUuid,
}: {
  containerUuid: string;
  shipUuid: string;
}) => getContainerAndShip({ containerUuid, shipUuid });

describe('withMutationEffects', () => {
  const exampleStore: Record<string, any> = {};
  const cacheGetMock = jest.fn(async (key) => exampleStore[key]);
  const cacheSetMock = jest.fn(
    async (key, value) => (exampleStore[key] = value),
  );
  const cache: SimpleAsyncCache<string> = {
    get: cacheGetMock,
    set: cacheSetMock,
  };
  const getContainerAndShipWithCaching = withQueryCaching(addContainerToShip, {
    cache,
    dependsOn: [
      // add a bunch of dependencies -> get a bunch of dependency pointers
      {
        identity: {
          dobj: ref(Container),
          uuid: ({ input }) => input[0].containerUuid,
        },
      },
      {
        relationship: {
          from: {
            dobj: ref(Ship),
            uuid: ({ input }) => input[0].shipUuid,
          },
          to: {
            dobj: ref(Container),
          },
          via: ref(Container, 'onShipUuid'),
        },
      },
      {
        relationship: {
          from: {
            dobj: ref(Container),
            uuid: ({ input }) => input[0].containerUuid,
          },
          to: {
            dobj: 'InsuredCargo',
          },
          via: ref(Container, 'insuredCargoUuids'),
        },
      },
    ],
  });

  const logDebugMock = jest.fn();
  const addContainerToShipWithMutationEffects = withMutationEffects(
    addContainerToShip,
    { cache, logDebug: logDebugMock },
  );

  beforeEach(() => jest.clearAllMocks());
  it('should check for queries to invalidate for all domain objects referenced in outputs', async () => {
    // run the mutation
    await addContainerToShipWithMutationEffects({
      containerUuid: '__container.uuid__',
      shipUuid: '__ship.uuid__',
    });

    // confirm we looked up queries to invalidate for each non-metadata property of the output referenced dobjs
    expect(cacheGetMock).toHaveBeenCalledTimes(17);
    expect(cacheGetMock.mock.calls).toMatchSnapshot(); // save an example
  });

  it('should produce query invalidations for all the domain entities referenced in outputs', async () => {
    const containerUuid = '__container.uuid__';
    const shipUuid = '__ship.uuid__';

    // get the container and ship with caching, to set some dependency pointers into the cache
    await getContainerAndShipWithCaching({
      containerUuid,
      shipUuid,
    });

    // now run the mutation
    await addContainerToShipWithMutationEffects({
      containerUuid,
      shipUuid,
    });

    // confirm we ran invalidations on all of the dependency pointers that existed
    expect(cacheSetMock).toHaveBeenCalledTimes(8);
    expect(cacheSetMock.mock.calls).toMatchSnapshot();
  });
  it('should log the cache.effects if logDebug option was specified', async () => {
    // run the mutation
    await addContainerToShipWithMutationEffects({
      containerUuid: '__container.uuid__',
      shipUuid: '__ship.uuid__',
    });

    // confirm we logged the effects
    expect(logDebugMock).toHaveBeenCalledTimes(1);
    expect(logDebugMock).toHaveBeenCalledWith('ddcache.mutation.effects', {
      updated: { references: expect.any(Array) },
      evaluated: { deps: expect.any(Array) },
      invalidated: { queries: expect.any(Array) },
    });
    expect(logDebugMock.mock.calls).toMatchSnapshot();
  });
});
