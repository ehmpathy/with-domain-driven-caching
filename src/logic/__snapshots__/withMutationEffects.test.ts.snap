// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withMutationEffects should check for queries to invalidate for all domain objects referenced in outputs 1`] = `
[
  [
    ".cache.ref.Container.__container.icid__.317f6449a26907227ea8de349c480310628297c300c6586e97ca8706f861c7d1",
  ],
  [
    ".cache.ref.Ship.__ship.isid__.0b689f94a3c7d857f5587801814a875598b4dfd25e9d836887a968556c9763f8",
  ],
  [
    ".query.dep.Container.uuid.__container.uuid__.icid",
  ],
  [
    ".query.dep.Container.icid.__containericid__.885f378972d5731706f1214eaf2c7227c17163034c03d0f1c93e1fd971079c48",
  ],
  [
    ".query.dep.Container.uuid.__container.uuid__.manufacturer",
  ],
  [
    ".query.dep.Container.manufacturer.containersrus.953bd78f87bfdefbe9c5745c823c503eb2ea3e64a1cfd388d17bbc422e06a9d6",
  ],
  [
    ".query.dep.Container.uuid.__container.uuid__.onShipUuid",
  ],
  [
    ".query.dep.Container.onShipUuid.__shipuuid__.11a185c00351c263db81a13d211b0fe9e2f039f30b467d8426e8bd1ea9e74a85",
  ],
  [
    ".query.dep.Container.uuid.__container.uuid__.insuredCargoUuids",
  ],
  [
    ".query.dep.Container.insuredCargoUuids.__insuredcargo1__.97c4311ea52323c110dc41035084a81560edf404337073990f2b9b790c388b7d",
  ],
  [
    ".query.dep.Container.insuredCargoUuids.__insuredcargo2__.44d32806d2699a360872669d123a5e813ecb8f3da642d9919bc857a0ec3fe66a",
  ],
  [
    ".query.dep.Container.uuid.__container.uuid__.lock",
  ],
  [
    ".query.dep.Container.lock.ContainerLock.swipernoswiping.lockymclockface.3cf448f9380ecd4daf668110f1f0bf7884de32b69c3576082bf0dcc1ca50c2b9",
  ],
  [
    ".query.dep.Ship.uuid.__ship.uuid__.isid",
  ],
  [
    ".query.dep.Ship.isid.__shipisid__.f905828ea72c78c5a27e62d09d6a7db443ba838326e8d24ad43766655561e500",
  ],
  [
    ".query.dep.Ship.uuid.__ship.uuid__.name",
  ],
  [
    ".query.dep.Ship.name.boatymcboatface.ade929700486ec399d87267309e214546b88038e99cd625adcdaba49ad5470bf",
  ],
]
`;

exports[`withMutationEffects should log the cache.effects if logDebug option was specified 1`] = `
[
  [
    "ddcache.mutation.effects",
    {
      "evaluated": {
        "deps": [
          ".query.dep.Container.uuid.__container.uuid__.lock",
          ".query.dep.Container.lock.manufacturerswipernoswipingmodellockymclockface.3cf448f9380ecd4daf668110f1f0bf7884de32b69c3576082bf0dcc1ca50c2b9",
          ".query.dep.Container.lock.ContainerLock.swipernoswiping.lockymclockface.3cf448f9380ecd4daf668110f1f0bf7884de32b69c3576082bf0dcc1ca50c2b9",
        ],
      },
      "invalidated": {
        "queries": [],
      },
      "updated": {
        "references": [
          ".cache.ref.Container.__container.icid__.317f6449a26907227ea8de349c480310628297c300c6586e97ca8706f861c7d1",
          ".cache.ref.Ship.__ship.isid__.0b689f94a3c7d857f5587801814a875598b4dfd25e9d836887a968556c9763f8",
        ],
      },
    },
  ],
]
`;

exports[`withMutationEffects should produce query invalidations for all the domain entities referenced in outputs 1`] = `
[
  [
    ".query.dep.Container.uuid.__containeruuid__.41a9e6b0c8ab54ff319a9a1551e51a0045cd2f91a5ab2476e4995e47d2fb1d6d",
    "{"queries":["[{\\"containerUuid\\":\\"__container.uuid__\\",\\"shipUuid\\":\\"__ship.uuid__\\"}]"]}",
    {
      "secondsUntilExpiration": Infinity,
    },
  ],
  [
    ".query.dep.Container.onShipUuid.__shipuuid__.11a185c00351c263db81a13d211b0fe9e2f039f30b467d8426e8bd1ea9e74a85",
    "{"queries":["[{\\"containerUuid\\":\\"__container.uuid__\\",\\"shipUuid\\":\\"__ship.uuid__\\"}]"]}",
    {
      "secondsUntilExpiration": Infinity,
    },
  ],
  [
    ".query.dep.Container.uuid.__container.uuid__.insuredCargoUuids",
    "{"queries":["[{\\"containerUuid\\":\\"__container.uuid__\\",\\"shipUuid\\":\\"__ship.uuid__\\"}]"]}",
    {
      "secondsUntilExpiration": Infinity,
    },
  ],
  [
    "[{"containerUuid":"__container.uuid__","shipUuid":"__ship.uuid__"}]",
    "{"container":".cache.ref.Container.__container.icid__.317f6449a26907227ea8de349c480310628297c300c6586e97ca8706f861c7d1","ship":".cache.ref.Ship.__ship.isid__.0b689f94a3c7d857f5587801814a875598b4dfd25e9d836887a968556c9763f8"}",
    {
      "secondsUntilExpiration": undefined,
    },
  ],
  [
    ".cache.ref.Container.__container.icid__.317f6449a26907227ea8de349c480310628297c300c6586e97ca8706f861c7d1",
    "{"uuid":"__container.uuid__","icid":"__container.icid__","manufacturer":"containers r us","onShipUuid":"__ship.uuid__","insuredCargoUuids":["__insured.cargo.1__","__insured.cargo.2__"],"lock":{"manufacturer":"swiper no swiping","model":"locky mclockface"}}",
    {
      "secondsUntilExpiration": 2592000,
    },
  ],
  [
    ".cache.ref.Ship.__ship.isid__.0b689f94a3c7d857f5587801814a875598b4dfd25e9d836887a968556c9763f8",
    "{"uuid":"__ship.uuid__","isid":"__ship.isid__","name":"boaty mcboatface"}",
    {
      "secondsUntilExpiration": 2592000,
    },
  ],
  [
    ".cache.ref.Container.__container.icid__.317f6449a26907227ea8de349c480310628297c300c6586e97ca8706f861c7d1",
    "{"uuid":"__container.uuid__","icid":"__container.icid__","manufacturer":"containers r us","onShipUuid":"__ship.uuid__","insuredCargoUuids":["__insured.cargo.1__","__insured.cargo.2__"],"lock":{"manufacturer":"swiper no swiping","model":"locky mclockface"}}",
    {
      "secondsUntilExpiration": 2592000,
    },
  ],
  [
    ".cache.ref.Ship.__ship.isid__.0b689f94a3c7d857f5587801814a875598b4dfd25e9d836887a968556c9763f8",
    "{"uuid":"__ship.uuid__","isid":"__ship.isid__","name":"boaty mcboatface"}",
    {
      "secondsUntilExpiration": 2592000,
    },
  ],
]
`;
